Filter notifications by content type

ioanaciont opened the pull request:
Clients can filter the content adding a request param for content type (e.g. https://api.ft.com/content/notifications-push?type=article)
For the moment, we do not support multiple types in a single request -> agreed with David L. (`type=article&type=content` does not work as expected, it will filter results based on the first value from the request -> article)

- If the type is not specified, the filter will be set to `type=Article`
- Clients can specify `type=All` that will return all the notifications, despite of their type.
- `DELETE` notifications will be always sent to the clients, because their payload is empty (in Kafka) and we cannot extract the type

Supported types in notifications push are: `Article`, `Content` (for content placeholders), `ContentPackage`.

davidbalazs93 commented on resources/push.go:
```diff
@@ -11,10 +11,13 @@ import (
+	defaultContentType = "Article"
 )
 
+var supportedContentTypes = []string{"Article", "Content", "ContentPackage", "All"}
```
it would be nice to have this array as a const one, agree? :)

ioanaciont replied:
@davidbalazs93 I was also thinking about this, but then I found that you can't declare constant arrays in Go. The explanation I found is that `Go consts are things that are evaluated at compile-time. Array values don't exist at compile-time.` 
More on this topic: https://stackoverflow.com/questions/13137463/declare-a-constant-array

kovacshuni commented on dispatcher/subscribers.go:
```diff
@@ -45,6 +49,17 @@ func (s *standardSubscriber) Since() time.Time {
+	if n.Type == "DELETE" || strings.ToLower(s.acceptedContentType) == "all" {
+		return true
+	}
+	if strings.ToLower(s.acceptedContentType) == strings.ToLower(contentType) {
```
Here it's enough if you return the equality condition, without the `if`, `true` and `false`.

ioanaciont replied:
@kovacshuni Yep you're right, I'm going to change it :)

kovacshuni commented:
I don't know why is not the `contentType` inside the `Notification` struct. We have multiple levels now. I know maybe you wanted to leave the `Notification` struct to remain compatible with the `/notifications` format, but i don't know if it's worth it. You could discuss with others, Next and PO's.

Very nicely written description (of the PR).

ioanaciont commented:
Thanks @kovacshuni! Is there a way to include `contentType` as a field to `Notification` struct, without exposing it in the json response? I've tried it, but I couldn't find anything useful on this topic. 

As far as I know we don't want to send `contentType` to the Push API clients, because we try to keep the messages quick & short. Also, adding this field might require additional processing at client side, and we would like to avoid it if possible.

kovacshuni commented on dispatcher/subscribers.go:
```diff
@@ -53,11 +53,8 @@ func (s *standardSubscriber) matchesContentType(n Notification, contentType stri
-	}
 
-	return false
+	return (strings.ToLower(s.acceptedContentType) == strings.ToLower(contentType))
```
no need for the outer `(`,`)`s :)

MihaiSorinMoisa commented on dispatcher/notifications.go:
```diff
@@ -12,6 +12,7 @@ type Notification struct {
 	Standout         Standout `json:"standout"`
 }
 
+// Standout model for a Notification
```
The constant `changeType` from this file is unused and should be removed.

MihaiSorinMoisa commented on dispatcher/notifications.go:
```diff
@@ -12,6 +12,7 @@ type Notification struct {
 }
 
+// Standout model for a Notification
 type Standout struct {
```
The constant `changeType` from this file is unused and should be removed.

bogdanguranda commented on resources/push.go:
```diff
@@ -11,10 +11,13 @@ import (
+	defaultContentType = "Article"
 )
 
+var supportedContentTypes = []string{"Article", "Content", "ContentPackage", "All"}
```
Do you know if we're sending notifications for ImageSet, MediaResource, Video, also? If we do, we should add them here also.

ioanaciont replied:
@bogdanguranda No, we don't send notifications for those. _Supported types in notifications push are: Article, Content (for content placeholders), ContentPackage._

bogdanguranda replied:
Ok, thank you, good to know :)

MihaiSorinMoisa commented on consumer/consume.go:
```diff
@@ -57,9 +57,11 @@ func (qHandler *simpleMessageQueueHandler) HandleMessage(queueMsg kafka.FTMessag
 		log.WithField("transaction_id", msg.TransactionID()).WithField("msg", string(msg.Body)).WithError(err).Warn("Skipping event: Cannot build notification for message.")
```
This flow should be tested as well.

MihaiSorinMoisa commented on consumer/mapper.go:
```diff
@@ -67,3 +67,17 @@ func extractDataFromPayload(event PublicationEvent) (string, bool) {
+		return ""
+	}
+
+	notificationPayloadMap, ok := event.Payload.(map[string]interface{})
```
The following 2 flows should be tested as well.

ioanaciont commented:
@davidbalazs93 @bogdanguranda @MihaiSorinMoisa @kovacshuni I moved `contentType` field inside the `Notification` struct, based on Huni's suggestion - I hope the code is a bit cleaner now. I also updated the implementation & the unit tests considering all the other suggestions. Thanks all for being involved, I really appreciate it :)

kovacshuni commented:
looking good

ioanaciont commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Avoid notifying push clients about content they can't use](https://trello.com/c/s1Fg6LxJ/38-avoid-notifying-push-clients-about-content-they-cant-use)
